File: docs/algorithms/code/algorithms/reverseWords.ts
function reverseWords(s: string): string {
  const str = s.trim();

  let res = "";
  let index = str.length - 1;
  while (index >= 0) {
    const word: string[] = [];

    while (index >= 0 && str[index] !== " ") {
      word.unshift(str[index] as string);
      index--;
    }

    res += word.join("");

    if (index > 0) {
      res += " ";
    }
    while (index >= 0 && str[index] === " ") {
      index--;
    }
  }

  return res;
}

File: docs/algorithms/code/algorithms/reverseWords.go
// 思路：指针遍历，每次取 left 到 right 之间的字符串(左开右闭)，插入到数组的头部，指针指向 right
// 复杂度分析： 时间复杂度 O(n) 空间复杂度 O(n)
import "strings"

func reverseWords(s string) string {
	s = strings.TrimSpace(s) // 去除首尾空格
	n := len(s)
	arr := []string{}

	index := 0

	for index < n {
		for index < n && s[index] == ' ' {
			index++
		}

		if index >= n {
			break
		}

		start, end := index, index
		for end < n && s[end] != ' ' {
			end++
		}

		arr = append([]string{s[start:end]}, arr...)
		index = end
	}

	return strings.Join(arr, " ")
}